### PR TITLE
fix(evidence): make evidence sync resilient to peer-routing race

### PR DESCRIPTION
**Pull request by lklimek**
## Problem

`TestReactorBroadcastEvidence_Committed` was flaky under load (~1.5–3% failure
rate at 12-way concurrency). Root cause: `PeerStatusUp` fires before the p2p
router finishes wiring the channel route to the new peer. The one-shot
`syncEvidence` goroutine sent all pool evidence immediately, the sends returned
`nil` but envelopes were silently dropped, and the peer permanently missed
evidence that pre-dated its connection. `broadcastEvidence` only fires on
*newly received* evidence, so there was no self-heal path.

## Fix

Convert the one-shot walk in `syncEvidence` to a ticker loop that re-sends
every `evidenceSyncInterval` (1 s) until the peer disconnects. Duplicates are
idempotent on the receiver side (`isPending` / `isCommitted` guards in
`handleEvidenceMessage`) — no protocol changes required.

### QA-001 — Deterministic regression guard (`TestSyncEvidenceRetries`)

Wires a recording `p2p.Channel` stub and asserts that the same evidence is sent
**more than once** across ticker ticks (at 20 ms test interval, 150 ms window).
A pre-fix one-shot goroutine would send exactly `numEv` items and exit, making
`count > numEv` **FAIL** — this is a true regression gate.

### QA-002 — Peer-flap lifecycle race

With a long-running goroutine (ticker loop), the window where a stale goroutine
and a new one coexist on Down→Up grew. Fixed by:

1. `PeerStatusDown`: delete the map entry **synchronously** (not left for the
   goroutine's deferred cleanup), so a fast Down→Up reconnect always starts
   fresh.
2. `syncEvidence` carries a unique identity token (`*struct{}`); its deferred
   cleanup only removes the map entry if the stored token still matches —
   prevents it from evicting a new goroutine's entry.

### QA-003/004 — Documentation & interval

- `doc.go`: removed stale `broadcastEvidenceIntervalS` / 60 s reference;
  documents both propagation paths (per-peer sync loop + broadcast on new
  evidence).
- `reactor.go`: `evidenceSyncInterval` changed from `const` to `var` (allows
  test override via `SetEvidenceSyncIntervalForTesting`), now 1 s with detailed
  comment on resend profile and amplification rationale (10 s was considered but
  causes a 10× test-duration regression).

## Before / After

| Scenario | Before | After |
|---|---|---|
| Flaky test failure rate (12-way concurrent) | ~1.5–3 % | 0 % (150/150 in CI) |
| `TestSyncEvidenceRetries` | *(not present)* | PASS — fails without ticker loop |
| Full `./internal/evidence/` suite time | ~6 s | ~6.5 s |

## Checklist

- [x] `go test -race -count=1 ./internal/evidence/` → PASS (6.5 s)
- [x] `gofmt -w -s` on all changed files
- [x] `golangci-lint` — no new issues in changed files
- [x] `TestSyncEvidenceRetries` fails on pre-fix code (regression gate verified)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

### Review follow-up (thepastaclaw round 3) — addressed in `fb86e098`

- **`uint64` sync-identity token** — the per-peer `syncEvidence` goroutine identity is now a monotonic `uint64` counter on the `Reactor` (incremented under the install mutex) instead of a `*int`, removing reliance on heap-address distinctness and the latent `*struct{}`-regression footgun.
- **`MaxBytes`-bounded resync** — each retry tick now sends `evpool.PendingEvidence(ConsensusParams.Evidence.MaxBytes)` rather than walking the full pending clist, capping per-tick gossip at the same byte budget the proposer uses. Side benefit: the `ValidateABCI` repair-and-store subpath now actually re-emits via the ticker (the old clist walk skipped it).

**Bundled behavior change — rebroadcast suppression on unverifiable evidence:** The suppression of rebroadcast on non-`ErrInvalidEvidence` errors in `handleEvidenceMessage` is an intentional bundled change. When `AddEvidence` fails with a plain (non-`ErrInvalidEvidence`) error — typically because the node is behind and lacks the block needed to verify the evidence — the reactor now returns without re-gossiping, since broadcasting evidence we cannot verify only amplifies it across the mesh without ever converging. Eventual delivery still holds: any evidence that enters the pending set is re-emitted by the per-peer `syncEvidence` ticker, including the `ValidateABCI` repair-and-store subpath (`verify.go:74-79`), which writes the repaired evidence to the pending store and is therefore picked up by the ticker's `PendingEvidence` read. Locked by `TestReactorNoRebroadcastOnNonInvalidEvidenceError`.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved evidence propagation so connected peers receive updates more reliably over time.
  * Prevented duplicate re-sharing of evidence when a submission can’t be processed for non-validation reasons.
  * Fixed a reconnect edge case where peer sync could be interrupted during quick disconnect/reconnect activity.
  * Updated tests and documentation to match the revised evidence syncing behavior.

### CODE REVIEW

**Comment by lklimek** [comment]
@thepastaclaw please review

**Comment by lklimek** [comment]
@thepastaclaw please review

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

**Comment by thepastaclaw** [comment]
✅ Review complete (commit 08abb4e4)

File: internal/evidence/reactor.go [hunk 1]
```diff
@@ -22,6 +23,52 @@ const (
...
+// The identity token solves the Down→Up flap race: when PeerStatusDown fires,
+// the handler deletes the map entry synchronously and calls cancel(). The
+// goroutine is still running until it notices the cancellation, and its
+// deferred cleanup compares its own token against whatever is in the map at
+// that point — if a fast Up event already installed a new goroutine, the
+// tokens differ and the old goroutine does not delete the new entry.
+type peerSyncState struct {
+	cancel context.CancelFunc
+	id     *struct{} // unique pointer per goroutine; comparable as a pointer
+}
```
Comment (thepastaclaw) [inline internal/evidence/reactor.go:81]: **🔴 Blocking: Identity token uses `*struct{}`, so all tokens alias to runtime.zerobase and the Down→Up flap guard never fires**

`peerSyncState.id` is typed `*struct{}` and constructed with `new(struct{})` (reactor.go:281). Per the Go spec: "Pointers to distinct zero-size variables may or may not be equal." In the current Go runtime, every zero-size allocation returns `&runtime.zerobase`, so every `new(struct{})` returns the same address. The `current.id == myID` check in the deferred cleanup (reactor.go:337) therefore always evaluates true regardless of which goroutine owns the entry.

Concrete failure path — the exact scenario the token was added to prevent:

1. PeerStatusUp → install entry M1 with token T1 (= &zerobase), start G1.
2. PeerStatusDown → cancel G1, delete M1 synchronously, release mtx.
3. G1's defer is preempted before grabbing mtx.
4. Fast PeerStatusUp arrives → installs M2 with token T2 (= &zerobase), starts G2.
5. G1's defer finally runs: `&zerobase == &zerobase` → true → G1 deletes M2 from the map.

Result: G2 keeps running with no map entry, so the next PeerStatusDown for that peer finds nothing to cancel and G2 leaks until the parent ctx ends. If the peer reconnects, PeerStatusUp sees an empty slot and spawns G3 → two ticker loops sending evidence concurrently for the same peer, accumulating over repeated flaps.

Neither new test exercises this flap path (TestSyncEvidenceRetries does not flap; TestReactorMultiDisconnect does not reconnect), so CI green is not evidence the guard works.

Fix: use a non-zero-size pointee, e.g. `*int` (or `*uint64`) with `new(int)` at the construction site and update `syncEvidence`'s parameter type accordingly. The comments at 60–70, 244–252 and 327–329 then correctly describe the behavior.

```suggestion
type peerSyncState struct {
	cancel context.CancelFunc
	// id is a unique pointer per goroutine, compared by pointer identity.
	// MUST NOT be a pointer to a zero-size type — Go's runtime allocates all
	// zero-size values at runtime.zerobase, so `new(struct{})` returns the
	// same address every time and the identity check would always pass.
	id *int
}
```

<sub>source: ['claude']</sub>

**Review by thepastaclaw — CHANGES_REQUESTED** [review]
## Code Review

The PR's core mechanism (1s re-sync ticker) is sound and the Down→Up flap reasoning is mostly correct, but the identity-token implementation defeats itself: `*struct{}` pointers all alias to `runtime.zerobase`, so the guard the comments describe never fires. A separate regression also appears at the receiver side — every tick now causes lagging peers that fail verify-block to rebroadcast the same evidence indefinitely, where previously this happened at most once per connection.

🔴 1 blocking | 🟡 1 suggestion(s)

_1 additional finding(s) omitted (not in diff)._

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `internal/evidence/reactor.go`:
- [BLOCKING] internal/evidence/reactor.go:67-70: Identity token uses `*struct{}`, so all tokens alias to runtime.zerobase and the Down→Up flap guard never fires
  `peerSyncState.id` is typed `*struct{}` and constructed with `new(struct{})` (reactor.go:281). Per the Go spec: "Pointers to distinct zero-size variables may or may not be equal." In the current Go runtime, every zero-size allocation returns `&runtime.zerobase`, so every `new(struct{})` returns the same address. The `current.id == myID` check in the deferred cleanup (reactor.go:337) therefore always evaluates true regardless of which goroutine owns the entry.

  Concrete failure path — the exact scenario the token was added to prevent:

  1. PeerStatusUp → install entry M1 with token T1 (= &zerobase), start G1.
  2. PeerStatusDown → cancel G1, delete M1 synchronously, release mtx.
  3. G1's defer is preempted before grabbing mtx.
  4. Fast PeerStatusUp arrives → installs M2 with token T2 (= &zerobase), starts G2.
  5. G1's defer finally runs: `&zerobase == &zerobase` → true → G1 deletes M2 from the map.

  Result: G2 keeps running with no map entry, so the next PeerStatusDown for that peer finds nothing to cancel and G2 leaks until the parent ctx ends. If the peer reconnects, PeerStatusUp sees an empty slot and spawns G3 → two ticker loops sending evidence concurrently for the same peer, accumulating over repeated flaps.

  Neither new test exercises this flap path (TestSyncEvidenceRetries does not flap; TestReactorMultiDisconnect does not reconnect), so CI green is not evidence the guard works.

  Fix: use a non-zero-size pointee, e.g. `*int` (or `*uint64`) with `new(int)` at the construction site and update `syncEvidence`'s parameter type accordingly. The comments at 60–70, 244–252 and 327–329 then correctly describe the behavior.
- [SUGGESTION] internal/evidence/reactor.go:174-185: Lagging peers rebroadcast evidence on every tick because handleEvidenceMessage ignores AddEvidence errors
  Receiver-side dedup via `isPending`/`isCommitted` does not engage when the peer is behind the evidence height. In that case `Pool.verify` returns a non-`ErrInvalidEvidence` error (`missing block for height %d` at internal/evidence/verify.go:37–40), `AddEvidence` does not stash the item in pending, and the code falls through and still calls `broadcastEvidence` — and on the next received copy, `isPending`/`isCommitted` are both still false, so the broadcast happens again.

  Pre-PR this was bounded: syncEvidence sent each pre-existing item exactly once per peer connection. Post-PR the new ticker resends the same items every 1 s for the entire pending lifetime, so a lagging peer will keep failing verification and keep rebroadcasting the same evidence to all of its peers every tick until it catches up or disconnects. That's a sustained duplicate-gossip / log-spam regression directly exposed by this PR's new behavior.

  Minimal fix: only broadcast when `AddEvidence` succeeds (the comment immediately above the branch — "If the evidence is already pending or committed, we don't need to broadcast it again" — already implies this was the intent).
```
</details>

File: internal/evidence/reactor.go (hunk 1 above)
Comment (lklimek) [reply internal/evidence/reactor.go:81]: Fixed in `4c9d5f0`. `peerSyncState.id` is now `*int` (`new(int)`), with `syncEvidence`'s `myID` param updated to match — distinct heap addresses per goroutine, so the `current.id == myID` cleanup guard fires correctly and a stale `defer` can no longer evict a newer goroutine's map entry. The reviewer's suggested `*int`/`new(int)` fix matches exactly.

Added `TestReactorPeerFlapNoGoroutineEviction`, which deterministically forces the Down→Up interleaving (freezes G1 mid-pool-walk, flaps Down→Up to install G2, then releases G1's `defer`) — it **FAILS on the pre-fix `*struct{}` token** (both tokens alias to `runtime.zerobase`, asserted via the token-distinctness check) and passes with `*int`.

The companion rebroadcast-on-`AddEvidence`-failure suggestion is also handled: a non-`ErrInvalidEvidence` error now logs and `return nil` instead of falling through to `broadcastEvidence`, so a lagging peer no longer re-gossips the same evidence every tick.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

**Comment by lklimek** [comment]
@thepastaclaw please re-review — the BLOCKING finding and the rebroadcast suggestion are addressed in `4c9d5f0`:
- `peerSyncState.id` is now `*int` (`new(int)`), `syncEvidence` param updated to match — non-zero-size so the Down→Up flap guard's identity check actually fires (your suggested `*int`/`new(int)` exactly);
- added `TestReactorPeerFlapNoGoroutineEviction`, deterministic, **fails on the pre-fix `*struct{}` token** / passes with `*int`;
- non-`ErrInvalidEvidence` `AddEvidence` errors now `return nil` instead of falling through to `broadcastEvidence` (no more every-tick rebroadcast for lagging peers).

gofmt/vet/golangci-lint clean; `-race ×3` green.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

**Comment by lklimek** [comment]
@thepastaclaw review please

File: internal/evidence/reactor.go [hunk 2]
```diff
@@ -131,7 +186,16 @@ func (r *Reactor) handleEvidenceMessage(ctx context.Context, envelope *p2p.Envel
...
+				// most commonly because we are behind and lack the block needed to
+				// verify it (verify() returns a plain error, NOT ErrInvalidEvidence,
+				// for a missing block so the peer is not disconnected). The evidence
+				// never entered the pending set, so isPending stays false. Falling
+				// through to broadcastEvidence here would re-gossip on every receipt
+				// without ever converging — sustained amplification across the
+				// network. Log and stop; only successfully-added evidence is broadcast.
 				logger.Error("failed to add evidence", "error", err)
+				return nil
 			}
```
Comment (thepastaclaw) [inline internal/evidence/reactor.go:206]: **🟡 Suggestion: Bundled behavior change: handleEvidenceMessage no longer rebroadcasts on non-ErrInvalidEvidence errors — call this out in the PR description**

Pre-fix, when AddEvidence returned a non-ErrInvalidEvidence error (most commonly because the node is behind and lacks the referenced block — verify() in pool/verify.go:39 returns a plain error in that case), the receiver logged and fell through to `broadcastEvidence`, re-gossiping the unverified item. The new `return nil` suppresses that rebroadcast entirely.

This is defensible on its own merits — gossiping unverified evidence is amplification without convergence — and the inline comment explains the rationale clearly. The issues are scope and propagation:

1. The PR title and description frame the change as the peer-routing race / ticker loop fix; this hunk is a separate semantic change worth its own mention so reviewers can evaluate the tradeoff explicitly.
2. During state-sync / fast-sync, lagging nodes used to opportunistically forward evidence they could not yet verify. Post-fix that opportunistic forwarding is gone, and propagation depends entirely on each holder's syncEvidence ticker eventually re-firing to every peer. The ticker loop guarantees eventual delivery so consensus safety is not at risk (DuplicateVoteEvidence is slashing-relevant, not safety-relevant), but the propagation surface is narrower.
3. Note that verify()'s ValidateABCI repair-and-store branch (verify.go:74-79) inserts the evidence into pending while returning a non-ErrInvalidEvidence error — in that subpath the local syncEvidence loop will re-emit it, so behavior is preserved.

At minimum, document this in the PR description. A focused unit test (AddEvidence returns non-ErrInvalidEvidence => broadcastEvidence not called) would lock the new contract in.

<sub>source: ['claude', 'codex']</sub>

File: internal/evidence/reactor.go [hunk 3]
```diff
@@ -272,37 +366,46 @@ func (r *Reactor) syncEvidence(ctx context.Context, peerID types.NodeID) {
...
-		}); err != nil {
-			return
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			next = next.Next()
 		}
```
Comment (thepastaclaw) [inline internal/evidence/reactor.go:411]: **🟡 Suggestion: Bound each retry tick by ConsensusParams.Evidence.MaxBytes instead of walking the full pool clist**

The retry loop walks `EvidenceFront()` to the end and sends every pending item on every tick. The author's inline comment correctly notes the worst-case per-node send rate is P × M messages/s, but it considers a single sender; in a fully-meshed N-node deployment that has all detected the same equivocation, every receiver becomes a sender once it admits the evidence, so steady-state cluster traffic is ≈ N × (N-1) × M envelopes/s while items remain pending. Pools are normally small and drain within a block, so the burst window is short — but the loop has no protocol-level cap.

The consensus protocol already defines an authoritative byte budget for evidence at any given height: `ConsensusParams.Evidence.MaxBytes`. The pool's own `PendingEvidence(maxBytes)` (pool.go:91) already honors it for proposal selection. Using the same bound on the per-peer retry path keeps the gossip channel symmetric with what the proposer would actually include, and turns this loop's worst case into a fixed-bytes-per-tick ceiling instead of a pool-size-dependent one.

Not a blocker (the receiver dedupes idempotently and the loop is functionally correct), but a small, low-risk hardening with no behavioral downside.

<sub>source: ['claude', 'codex']</sub>

File: internal/evidence/reactor.go [hunk 4]
```diff
@@ -22,6 +23,60 @@ const (
...
+// zero-size allocation (e.g. new(struct{})) at the single shared
+// runtime.zerobase address, so distinct zero-size pointers compare EQUAL. That
+// would make every token identical and silently defeat the flap guard, letting
+// a stale goroutine evict a newer goroutine's map entry (goroutine leak / two
+// ticker loops per peer). A *int is non-zero-size, so each allocation is
+// guaranteed a unique address — exactly the identity this guard relies on.
+type peerSyncState struct {
+	cancel context.CancelFunc
+	id     *int // unique per goroutine; MUST be non-zero-size so addresses differ
+}
```
Comment (thepastaclaw) [inline internal/evidence/reactor.go:81]: **💬 Nitpick: Replace the `*int` identity token with a monotonic counter on the Reactor**

The token's correctness hinges on a Go runtime/allocator detail: each `new(int)` returns a distinct heap address, while `new(struct{})` would collapse to `runtime.zerobase`. The 7-line doc comment captures this well, but it is a fragile foundation — a well-meaning cleanup that switches the type back to `*struct{}`, or a future escape-analysis change, silently reintroduces the very flap-race this PR fixes.

A `uint64` counter on the Reactor, incremented under `r.mtx` (already held during install) and copied into `peerSyncState{id: nextID}`, would (a) be impossible to spoof regardless of value type, (b) be self-documenting (the field is plainly an identity counter), (c) replace `reflect.ValueOf(entry.id).Pointer()` in export_test.go with a plain integer compare, and (d) eliminate one heap allocation per Up event.

Current code is correct — this is pure hardening / readability.

<sub>source: ['claude']</sub>

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

PR #1366 cleanly fixes the peer-routing race in the evidence reactor by converting the one-shot pool walk into a 1s ticker loop and guards Down→Up flaps with a per-goroutine identity token; lifecycle, locking, and idempotency look correct, with two deterministic regression tests added. Worth raising before merge: (a) a quietly bundled behavior change in handleEvidenceMessage that suppresses rebroadcast on non-ErrInvalidEvidence failures, (b) the retry loop is unbounded by the consensus evidence byte budget so a large pending pool can amplify cluster-wide traffic, and (c) the new long-running per-peer goroutine relies entirely on parent-ctx cancellation (OnStop does not cancel it explicitly), which is the existing reactor pattern but is now more load-bearing.

🟡 3 suggestion(s) | 💬 1 nitpick(s)

_1 additional finding(s) omitted (not in diff)._

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `internal/evidence/reactor.go`:
- [SUGGESTION] internal/evidence/reactor.go:189-199: Bundled behavior change: handleEvidenceMessage no longer rebroadcasts on non-ErrInvalidEvidence errors — call this out in the PR description
  Pre-fix, when AddEvidence returned a non-ErrInvalidEvidence error (most commonly because the node is behind and lacks the referenced block — verify() in pool/verify.go:39 returns a plain error in that case), the receiver logged and fell through to `broadcastEvidence`, re-gossiping the unverified item. The new `return nil` suppresses that rebroadcast entirely.

  This is defensible on its own merits — gossiping unverified evidence is amplification without convergence — and the inline comment explains the rationale clearly. The issues are scope and propagation:

  1. The PR title and description frame the change as the peer-routing race / ticker loop fix; this hunk is a separate semantic change worth its own mention so reviewers can evaluate the tradeoff explicitly.
  2. During state-sync / fast-sync, lagging nodes used to opportunistically forward evidence they could not yet verify. Post-fix that opportunistic forwarding is gone, and propagation depends entirely on each holder's syncEvidence ticker eventually re-firing to every peer. The ticker loop guarantees eventual delivery so consensus safety is not at risk (DuplicateVoteEvidence is slashing-relevant, not safety-relevant), but the propagation surface is narrower.
  3. Note that verify()'s ValidateABCI repair-and-store branch (verify.go:74-79) inserts the evidence into pending while returning a non-ErrInvalidEvidence error — in that subpath the local syncEvidence loop will re-emit it, so behavior is preserved.

  At minimum, document this in the PR description. A focused unit test (AddEvidence returns non-ErrInvalidEvidence => broadcastEvidence not called) would lock the new contract in.
- [SUGGESTION] internal/evidence/reactor.go:372-399: Bound each retry tick by ConsensusParams.Evidence.MaxBytes instead of walking the full pool clist
  The retry loop walks `EvidenceFront()` to the end and sends every pending item on every tick. The author's inline comment correctly notes the worst-case per-node send rate is P × M messages/s, but it considers a single sender; in a fully-meshed N-node deployment that has all detected the same equivocation, every receiver becomes a sender once it admits the evidence, so steady-state cluster traffic is ≈ N × (N-1) × M envelopes/s while items remain pending. Pools are normally small and drain within a block, so the burst window is short — but the loop has no protocol-level cap.

  The consensus protocol already defines an authoritative byte budget for evidence at any given height: `ConsensusParams.Evidence.MaxBytes`. The pool's own `PendingEvidence(maxBytes)` (pool.go:91) already honors it for proposal selection. Using the same bound on the per-peer retry path keeps the gossip channel symmetric with what the proposer would actually include, and turns this loop's worst case into a fixed-bytes-per-tick ceiling instead of a pool-size-dependent one.

  Not a blocker (the receiver dedupes idempotently and the loop is functionally correct), but a small, low-risk hardening with no behavioral downside.
- [SUGGESTION] internal/evidence/reactor.go:146-148: syncEvidence's long-running loop now relies entirely on parent-ctx cancellation; OnStop should cancel it explicitly
  The loop exits only via `ctx.Done()` or a `Send` error. That ctx is the OnStart ctx, which is NOT canceled by `BaseService.Stop()` — `Stop()` calls `OnStop()` (which only closes the evpool) and cancels an internal `srvCtx` that is never passed to anywhere in the reactor. In normal node shutdown the parent ctx is canceled along with shutdown so the goroutine exits cleanly, and the existing `if !r.IsRunning()` guard on PeerStatusUp prevents new goroutines from being spawned after Stop — so this is not a production leak today.

  However, pre-PR `syncEvidence` was a one-shot pool walk that returned in milliseconds, so calling `Stop()` without also canceling the parent ctx was harmless. Post-PR the goroutine ticks forever in that scenario, which makes a previously-benign pattern observable as a goroutine leak (most likely surfaced first in tests that call `reactor.Stop()` directly).

  Cheapest fix: track a cancel func on the Reactor and call it from `OnStop()`, or store all per-peer cancels and invoke them on stop. This also documents the lifecycle contract explicitly rather than relying on every caller to remember to cancel the OnStart ctx.
```
</details>

File: internal/evidence/reactor.go (hunk 4 above)
Comment (lklimek) [reply internal/evidence/reactor.go:81]: Done in `fb86e098` (pushed in `08abb4e4`). Replaced the `*int` token with a monotonic `uint64` counter: added `Reactor.nextSyncID`, incremented under the same mutex that installs `peerSyncState` (ids start at 1, so 0 stays a clean "absent" sentinel). `peerSyncState.id` and `syncEvidence`'s `myID` are now `uint64`; the cleanup guard is a plain `current.id == myID` compare. `export_test.go` drops the `reflect`-based pointer accessor for a `uint64` one, and `TestReactorPeerFlapNoGoroutineEviction` now asserts the two goroutines get distinct ids (negative control: pinning the id to a constant fails the flap test). Eliminates the per-Up heap alloc and the `*struct{}`-regression footgun you flagged. Resolving.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

File: internal/evidence/reactor.go (hunk 3 above)
Comment (lklimek) [reply internal/evidence/reactor.go:411]: Done in `fb86e098`. `syncEvidence` now bounds each tick by `ConsensusParams.Evidence.MaxBytes` via `evpool.PendingEvidence(maxBytes)` instead of walking `EvidenceFront()`→end. The budget is reachable from the reactor through `evpool.State().ConsensusParams.Evidence.MaxBytes` — the same authoritative source the proposer uses. Per-item send-error return and the ctx/ticker exit are preserved. Bonus correctness you may not have intended but is real: `PendingEvidence` reads the pending **store**, whereas the old `EvidenceFront()` clist walk skipped the `ValidateABCI` repair-and-store subpath (verify.go:74-79) entirely — so repaired evidence now actually re-emits via the ticker, making the "eventual delivery" guarantee genuinely hold. Resolving.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

File: internal/evidence/reactor.go (hunk 2 above)
Comment (lklimek) [reply internal/evidence/reactor.go:206]: Done in `fb86e098`, and documented in the PR description. Added `TestReactorNoRebroadcastOnNonInvalidEvidenceError`: a real `Pool` with a blockStore whose `LoadBlockMeta` returns nil → `verify()` fails with a plain (non-`ErrInvalidEvidence`) "missing block" error → the test asserts no rebroadcast (zero sends on a recording channel), no peer disconnect, and `pool.Size()==0`. Negative control: reverting to the old fall-through-`broadcastEvidence` makes the test fail, so it's a real contract guard. The PR body now calls out the bundled behavior change explicitly (incl. the `ValidateABCI` repair-and-store subpath still re-emitting via the ticker, so eventual delivery holds). Resolving.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

**Comment by lklimek** [comment]
@thepastaclaw please re-review — all three round-3 items are addressed in `08abb4e4` (which also includes a fresh `v1.6-dev` merge, so #1365's catch-up flake fix is now in the branch):

- **`uint64` monotonic sync-identity token** replacing the `*int` — no more reliance on heap-address distinctness; the `*struct{}`-regression footgun is gone (flap test asserts distinct ids).
- **Per-tick resync bounded by `ConsensusParams.Evidence.MaxBytes`** via `evpool.PendingEvidence(maxBytes)` instead of the full `EvidenceFront()` walk (also fixes the `ValidateABCI` repaired-evidence re-emit gap).
- **`TestReactorNoRebroadcastOnNonInvalidEvidenceError`** locks the rebroadcast-suppression contract (negative control verified); the bundled behavior change is now documented in the PR description.

`go build ./...` + `go test -race ./internal/evidence/...` green; gofmt/govet/misspell clean. All three review threads resolved.

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

**Comment by lklimek** [comment]
@coderabbitai review all

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

PR #1366 correctly hardens the evidence reactor's per-peer sync against the p2p routing race: the monotonic uint64 sync identity prevents flap eviction, PendingEvidence(MaxBytes) keeps resync symmetric with proposer selection, and non-ErrInvalidEvidence AddEvidence failures no longer rebroadcast unverifiable evidence. Both convergent codex findings on OnStop lifecycle are technically real but pre-existing in pattern; the PR amplifies (does not introduce) the dependency on parent ctx cancellation. Keeping one suggestion-level finding for that PR-driven amplification.

🟡 1 suggestion(s)

_1 additional finding(s) omitted (not in diff)._

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `internal/evidence/reactor.go`:
- [SUGGESTION] internal/evidence/reactor.go:153-155: OnStop does not cancel per-peer syncEvidence goroutines
  BaseService.Stop() only cancels its own internal srvCtx (libs/service/service.go:128-170), not the ctx passed to OnStart. The per-peer pctx at reactor.go:304 is derived from the OnStart ctx, so Reactor.Stop() leaves the new long-running ticker loops running until the caller separately cancels the parent ctx. Before this PR syncEvidence was a one-shot drain that exited quickly; the ticker loop never exits on its own, so the lifetime amplification turns a previously self-resolving goroutine into an indefinitely live one calling PendingEvidence/Send against a closed pool. In production the node cancels ctx together with Stop so this is not a current runtime hazard, but tests and embedding harnesses that stop the reactor directly will now leak per-peer goroutines. The minimal fix is to walk peerRoutines under mtx in OnStop and call cancel() on each entry before closing the pool.
```
</details>
